File: epi_judge_python_solutions/number_of_traversals_matrix.py
import functools
import itertools

from test_framework import generic_test


def number_of_ways(n: int, m: int) -> int:
    @functools.lru_cache(None)
    def compute_number_of_ways_to_xy(x, y):
        if x == y == 0:
            return 1

        ways_top = 0 if x == 0 else compute_number_of_ways_to_xy(x - 1, y)
        ways_left = 0 if y == 0 else compute_number_of_ways_to_xy(x, y - 1)
        return ways_top + ways_left

    return compute_number_of_ways_to_xy(n - 1, m - 1)


def number_of_ways_space_efficient(n, m):
    if n < m:
        n, m = m, n

    table = [1] * m
    for i in range(1, n):
        prev_res = 0
        for j in range(m):
            table[j] += prev_res
            prev_res = table[j]
    return table[m - 1]


# Pythonic implementation of space efficient solution.
def number_of_ways_pythonic(n, m):
    if n < m:
        n, m = m, n

    table = [1] * m
    for _ in range(1, n):
        table = list(itertools.accumulate(table))
    return table[-1]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('number_of_traversals_matrix.py',
                                       'number_of_traversals_matrix.tsv',
                                       number_of_ways))

File: epi_judge_cpp_solutions/number_of_traversals_matrix.cc
#include <algorithm>
#include <memory>
#include <vector>

#include "test_framework/generic_test.h"

using std::make_unique;
using std::swap;
using std::vector;

int ComputeNumberOfWaysToXY(int, int, vector<vector<int>>*);

int NumberOfWays(int n, int m) {
  return ComputeNumberOfWaysToXY(
      n - 1, m - 1,
      make_unique<vector<vector<int>>>(n, vector<int>(m, 0)).get());
}

int ComputeNumberOfWaysToXY(int x, int y,
                            vector<vector<int>>* number_of_ways_ptr) {
  if (x == 0 && y == 0) {
    return 1;
  }

  vector<vector<int>>& number_of_ways = *number_of_ways_ptr;
  if (number_of_ways[x][y] == 0) {
    int ways_top =
        x == 0 ? 0 : ComputeNumberOfWaysToXY(x - 1, y, number_of_ways_ptr);
    int ways_left =
        y == 0 ? 0 : ComputeNumberOfWaysToXY(x, y - 1, number_of_ways_ptr);
    number_of_ways[x][y] = ways_top + ways_left;
  }
  return number_of_ways[x][y];
}

int NumberOfWaysSpaceEfficient(int n, int m) {
  if (n < m) {
    swap(n, m);
  }

  vector<int> table(m, 1);
  for (int i = 1; i < n; ++i) {
    int prev_res = 0;
    if (n < m) {
      swap(n, m);
    }
    for (int j = 0; j < m; ++j) {
      table[j] = table[j] + prev_res;
      prev_res = table[j];
    }
  }
  return table[m - 1];
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"n", "m"};
  return GenericTestMain(args, "number_of_traversals_matrix.cc",
                         "number_of_traversals_matrix.tsv", &NumberOfWays,
                         DefaultComparator{}, param_names);
}
